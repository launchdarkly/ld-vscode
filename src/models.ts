export class Resource {
	name: string;
	key: string;
	tags: Array<string>;
	_version: number;
}

export class Project extends Resource {
	environments: Array<Environment>;
}

export class Environment extends Resource {
	apiKey: string;
	version: number;
	_site: { href: string };
}

export class Flag extends Resource {
	environments: Map<String, Environment>;

	constructor(init?: Partial<Flag>) {
		super();
		Object.assign(this, init);
	}

	environmentVersion(env: string): number {
		if (!this.environments[env]) {
			return -1;
		}
		return this._version + this.environments[env].version;
	}
}

export class FlagConfiguration {
	key: string;
	variations: Array<any>;
	offVariation: any;
	fallthrough: any;
	prerequisites: any;
	targets: any;
	rules: Array<any>;
	on: boolean;
	version: number;
}

export class FlagWithConfiguration {
	flag: Flag;
	config: FlagConfiguration;
}
