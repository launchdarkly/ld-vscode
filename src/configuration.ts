import { WorkspaceConfiguration, workspace, ExtensionContext } from 'vscode';

const package_json = require('../package.json');

const DEFAULT_BASE_URI = 'https://app.launchdarkly.com';
const DEFAULT_STREAM_URI = 'https://stream.launchdarkly.com';

export class Configuration {
	private readonly ctx: ExtensionContext;
	accessToken = '';
	sdkKey = '';
	project = '';
	env = '';
	enableHover = true;
	enableAutocomplete = true;
	baseUri = DEFAULT_BASE_URI;
	streamUri = DEFAULT_STREAM_URI;

	constructor(ctx: ExtensionContext) {
		this.ctx = ctx;
		this.reload();
	}

	reload() {
		const config = workspace.getConfiguration('launchdarkly');
		for (const option in this) {
			if (option === 'ctx') {
				continue;
			}
			this[option] = config.get(option);
		}

		// If accessToken is configured in state, use it. Otherwise, fall back to the legacy access token.
		this.accessToken = this.getState('accessToken') || this.accessToken;
	}

	async update(key: string, value: string | boolean, global: boolean) {
		if (typeof this[key] !== typeof value) {
			return;
		}

		let config: WorkspaceConfiguration = workspace.getConfiguration('launchdarkly');
		if (key === 'accessToken') {
			const ctxState = global ? this.ctx.globalState : this.ctx.workspaceState;
			await ctxState.update(key, value);
			await config.update(key, '', global);
			return;
		}

		await config.update(key, value, global);
		config = workspace.getConfiguration('launchdarkly');

		this[key] = value;
		process.nextTick(function() {});
	}

	validate(): string {
		const version = package_json.version;
		const ctx = this.ctx;
		ctx.globalState.update('version', undefined);
		const storedVersion = ctx.globalState.get('version');

		if (version !== storedVersion) {
			ctx.globalState.update('version', version);
		}

		const legacyConfiguration = !!this.sdkKey;
		if (legacyConfiguration && !ctx.globalState.get('legacyNotificationDismissed')) {
			return 'legacy';
		}

		// Only recommend configuring the extension on install and update
		const configured = !!this.accessToken;
		if (version != storedVersion && !configured) {
			return 'unconfigured';
		}
	}

	getState(key: string): string {
		return this.ctx.workspaceState.get(key) || this.ctx.globalState.get(key);
	}
}
