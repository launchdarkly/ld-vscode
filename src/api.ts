import * as rp from 'request-promise-native';
import * as url from 'url';

import { Configuration } from './configuration';
import { Resource, Project, Environment, Flag } from './models';

// LaunchDarklyAPI is a wrapper around request-promise-native for requesting data from LaunchDarkly's REST API. The caller is expected to catch all exceptions.
export class LaunchDarklyAPI {
	private readonly config: Configuration;

	constructor(config: Configuration) {
		this.config = config;
	}

	async getAccount() {
		const options = this.createOptions('account');
		const account = await rp(options);
		return JSON.parse(account);
	}

	async getProjects(): Promise<Array<Project>> {
		const options = this.createOptions('projects');
		const data = await rp(options);
		const projects = JSON.parse(data).items;
		projects.forEach((proj: Project) => {
			proj.environments = proj.environments.sort(sortNameCaseInsensitive);
			return proj;
		});
		return projects.sort(sortNameCaseInsensitive);
	}

	async getEnvironment(projectKey: string, envKey: string): Promise<Environment> {
		const options = this.createOptions(`projects/${projectKey}/environments/${envKey}`);
		const data = await rp(options);
		return JSON.parse(data);
	}

	async getFeatureFlag(projectKey: string, flagKey: string, envKey?: string): Promise<Flag> {
		const envParam = envKey ? '?env=' + envKey : '';
		const options = this.createOptions(`flags/${projectKey}/${flagKey + envParam}`);
		const data = await rp(options);
		return JSON.parse(data);
	}

	private createOptions(path: string) {
		return {
			url: url.resolve(this.config.baseUri, `api/v2/${path}`),
			headers: {
				Authorization: this.config.accessToken,
			},
		};
	}
}

const sortNameCaseInsensitive = (a: Resource, b: Resource) => {
	return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
};
