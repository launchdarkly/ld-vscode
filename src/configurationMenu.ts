import { QuickPickItem, window, workspace } from 'vscode';

import { MultiStepInput } from './multiStepInput';
import { LaunchDarklyAPI } from './api';
import { Resource, Project } from './models';
import { Configuration } from './configuration';

export class ConfigurationMenu {
	private readonly config: Configuration;
	private api: LaunchDarklyAPI;
	private title: string;
	private totalSteps: number;
	private currentAccessToken: string;
	private accessToken: string;
	private projects: Array<Project>;
	private project: string;
	private env: string;
	private useGlobalState: boolean;
	private invalidAccessToken: string;

	constructor(config: Configuration, api: LaunchDarklyAPI) {
		this.config = config;
		this.api = api;
		this.title = 'Configure LaunchDarkly';
		this.totalSteps = 3;
		workspace.name && this.totalSteps++;
		this.currentAccessToken = config.accessToken;
	}

	async collectInputs() {
		if (this.currentAccessToken) {
			await MultiStepInput.run(input => this.pickCurrentOrNewAccessToken(input));
			return;
		}

		await MultiStepInput.run(input => this.inputAccessToken(input));
	}

	shouldResume() {
		// Required by multiStepInput
		// Could show a notification with the option to resume.
		return new Promise<boolean>(() => {});
	}

	async pickCurrentOrNewAccessToken(input: MultiStepInput) {
		const existingTokenName = 'Use the existing access token';
		const options = [
			{ name: existingTokenName, key: 'xxxx' + this.currentAccessToken.substr(this.currentAccessToken.length - 6) },
			{ name: 'Enter a new access token' },
		].map(this.createQuickPickItem);

		const pick = await input.showQuickPick({
			title: this.title,
			step: 1,
			totalSteps: this.totalSteps,
			placeholder: 'Use your existing LaunchDarkly access token, or enter a new one.',
			items: options,
			shouldResume: this.shouldResume,
		});

		if (pick.label === existingTokenName) {
			this.accessToken = this.currentAccessToken;
			this.invalidAccessToken = '';
			return (input: MultiStepInput) => this.pickProject(input);
		}
		return (input: MultiStepInput) => this.inputAccessToken(input);
	}

	async inputAccessToken(input: MultiStepInput) {
		this.accessToken = '';
		this.accessToken = await input.showInputBox({
			title: this.title,
			step: 1,
			totalSteps: this.totalSteps,
			value: typeof this.accessToken === 'string' ? this.accessToken : '',
			prompt: 'Enter your LaunchDarkly access token',
			validate: token => this.validateAccessToken(token, this.invalidAccessToken),
			shouldResume: this.shouldResume,
		});

		try {
			this.updateAPI();
			await this.api.getAccount();
			return (input: MultiStepInput) => this.pickProject(input);
		} catch (err) {
			if (err.statusCode === 401) {
				this.invalidAccessToken = this.accessToken;
				window.showErrorMessage('Invalid access token, please try again.');
				return (input: MultiStepInput) => this.inputAccessToken(input);
			}
			throw err;
		}
	}

	async pickProject(input: MultiStepInput) {
		let projectOptions: QuickPickItem[];
		try {
			this.updateAPI();
			const projects = await this.api.getProjects();
			this.projects = projects;
			projectOptions = projects.map(this.createQuickPickItem);
		} catch (err) {
			if (err.statusCode === 401) {
				this.invalidAccessToken = this.accessToken;
				window.showErrorMessage('Invalid access token, please reconfigure your access token.');
				return (input: MultiStepInput) => this.inputAccessToken(input);
			}
			throw err;
		}

		const pick = await input.showQuickPick({
			title: this.title,
			step: 2,
			totalSteps: this.totalSteps,
			placeholder: 'Select a project',
			items: projectOptions,
			activeItem: typeof this.project !== 'string' ? this.project : undefined,
			shouldResume: this.shouldResume,
		});

		this.project = pick.description;
		return (input: MultiStepInput) => this.pickEnvironment(input);
	}

	async pickEnvironment(input: MultiStepInput) {
		const selectedProject = this.projects.find(proj => proj.key === this.project);
		const environments = selectedProject.environments;
		const environmentOptions = environments.map(this.createQuickPickItem);

		const pick = await input.showQuickPick({
			title: this.title,
			step: 3,
			totalSteps: this.totalSteps,
			placeholder: 'Select an environment',
			items: environmentOptions,
			activeItem: typeof this.env !== 'string' ? this.env : undefined,
			shouldResume: this.shouldResume,
		});

		this.env = pick.description;

		if (workspace.name) {
			return (input: MultiStepInput) => this.pickStorageType(input);
		}

		this.useGlobalState = true;
	}

	async pickStorageType(input: MultiStepInput) {
		const allWorkspacesName = 'All workspaces';
		const storageOptions = [
			{ name: 'This workspace', key: workspace.name },
			{ name: allWorkspacesName, key: 'Workspace-specific configurations will take precedence' },
		].map(this.createQuickPickItem);

		const pick = await input.showQuickPick({
			title: this.title,
			step: 4,
			totalSteps: this.totalSteps,
			placeholder: 'Pick a configuration type',
			items: storageOptions,
			shouldResume: this.shouldResume,
		});
		this.useGlobalState = pick.label == allWorkspacesName;
	}

	async validateAccessToken(token: string, invalidAccessToken: string) {
		if (token === invalidAccessToken) {
			return 'Invalid access token.';
		}
	}

	didChangeAccessToken(): boolean {
		return this.accessToken !== this.currentAccessToken;
	}

	updateAPI() {
		const configWithUpdatedToken = Object.assign({}, this.config);
		configWithUpdatedToken.accessToken = this.accessToken;
		this.api = new LaunchDarklyAPI(configWithUpdatedToken);
	}

	async configure() {
		await this.collectInputs();
		['accessToken', 'project', 'env'].forEach(async option => {
			await this.config.update(option, this[option], this.useGlobalState);
		});
		await this.config.update('sdkKey', '', true);
		if (workspace.name) {
			await this.config.update('sdkKey', '', false);
		}
	}

	createQuickPickItem(resource: Resource): QuickPickItem {
		return {
			label: resource.name,
			description: resource.key,
			detail: resource.tags && resource.tags.length > 0 && `Tags: ${resource.tags.join(', ')}`,
		};
	}
}
