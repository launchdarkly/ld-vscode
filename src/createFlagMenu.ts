import { env, window, workspace } from 'vscode';

import { MultiStepInput } from './multiStepInput';
import { LaunchDarklyAPI } from './api';
import { Configuration } from './configuration';
import { kebabCase } from 'lodash';
import { NewFlag } from './models';
export interface State {
	name: string;
	key: string;
	description: string;
	tags: string[];
	kind: string;
	clientAvailability: string;
	mobileAvailability: string;
	temporary: boolean;
}

/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
export class CreateFlagMenu {
	private readonly config: Configuration;
	private api: LaunchDarklyAPI;
	private title: string;
	private totalSteps: number;

	constructor(config: Configuration, api: LaunchDarklyAPI) {
		this.config = config;
		this.api = api;
		this.title = 'Create Feature Flag';
		this.totalSteps = 2;
		workspace.name && this.totalSteps++;
	}

	async collectInputs() {
		const state = {} as Partial<State>;
		await MultiStepInput.run((input) => this.setFlagName(input, state));
	}

	shouldResume(): Promise<boolean> {
		// Required by multiStepInput
		// Could show a notification with the option to resume.
		//eslint-disable-next-line @typescript-eslint/no-empty-function
		return new Promise<boolean>(() => {});
	}

	async setFlagName(input: MultiStepInput, state: Partial<State>) {
		const name = await input.showInputBox({
			title: this.title,
			step: 1,
			value: '',
			prompt: 'Enter Flag Name',
			totalSteps: this.totalSteps,
			shouldResume: this.shouldResume,
			validate: (token) => isValidName(token),
		});
		state.name = name;
		return (input: MultiStepInput) => this.setFlagKey(input, state);
	}

	async setFlagKey(input: MultiStepInput, state: Partial<State>) {
		const key = await input.showInputBox({
			title: this.title,
			step: 2,
			value: convertToKey(state.name),
			prompt: 'Enter Flag Key',
			totalSteps: this.totalSteps,
			shouldResume: this.shouldResume,
			validate: (token) => isValidKey(token),
		});
		state.key = key;
		return (input: MultiStepInput) => this.setAvailability(input, state);
	}

	async setAvailability(input: MultiStepInput, state: Partial<State>) {
		const enableServer = 'Only make flag available on Server';
		const enableClient = 'Enable Client Side';
		const enableMobile = 'Enable Mobile Side';
		const enableBoth = 'Enable Both Mobile and Client';

		const availability = await input.showQuickPick({
			title: this.title,
			step: 3,
			items: [{ label: enableServer }, { label: enableClient }, { label: enableMobile }, { label: enableBoth }],
			placeholder: 'Select Client-side Availability',
			totalSteps: this.totalSteps,
			shouldResume: this.shouldResume,
		});

		const flagAvailability = {
			usingEnvironmentId: false,
			usingMobileKey: false,
		};
		switch (availability.label) {
			case enableClient:
				flagAvailability.usingEnvironmentId = true;
				break;
			case enableMobile:
				flagAvailability.usingMobileKey = true;
				break;
			case enableBoth:
				flagAvailability.usingEnvironmentId = true;
				flagAvailability.usingMobileKey = true;
				break;
			default:
				break;
		}

		const buildFlag: NewFlag = Object.assign(state);
		buildFlag['clientSideAvailability'] = flagAvailability;
		// Remove empty fields
		Object.keys(buildFlag).forEach((k) => buildFlag[k] == null && delete buildFlag[k]);

		try {
			const flag = await this.api.postFeatureFlag(this.config.project, buildFlag);
			env.clipboard.writeText(flag.key);
			window.showInformationMessage(`Flag: ${flag.key} created and key copied to your clipboard.`);
		} catch (err) {
			window.showErrorMessage(`[LaunchDarkly] Creating flag ${err}`);
		}
	}
}

const keyRegexp = /^[\w\d][.A-Za-z_\-0-9]*$/u;
const capitalizedWordRegexp = /^[A-Z0-9][a-z0-9]*$/;
const tagRegexp = /^[.A-Za-z_\-0-9]+$/;

export const convertToKey = (input: string) => {
	if (!keyRegexp.test(input)) {
		return kebabCase(input);
	} else if (capitalizedWordRegexp.test(input)) {
		return input.toLowerCase();
	}
	return input;
};

const isValidKey = async (v: string): Promise<string> => {
	if (keyRegexp.test(v)) {
		return '';
	} else {
		return "Invalid key. Keys must start with a letter or number and only contain letters, numbers, '.', '_' or '-'";
	}
};

// TODO: Implement valid name check
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const isValidName = async (v: string): Promise<string> => {
	return '';
};

const isValidTags = async (v: string): Promise<string> => {
	const tags = v.split(',').filter((i) => i);
	for (const tag of tags) {
		if (!tagRegexp.test(tag)) {
			return `Tag: ${tag} is invalid it must contain only letters, number, '.', '_' or '-'`;
		}
	}

	return '';
};
