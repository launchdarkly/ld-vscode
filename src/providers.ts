import {
	commands,
	languages,
	window,
	CompletionItem,
	CompletionItemKind,
	CompletionItemProvider,
	DocumentFilter,
	ExtensionContext,
	Hover,
	HoverProvider,
	Position,
	Range,
	TextDocument,
} from 'vscode';
import * as url from 'url';
import opn = require('opn');
import { kebabCase } from 'lodash';

import { Configuration } from './configuration';
import { ConfigurationMenu } from './configurationMenu';
import { LaunchDarklyAPI } from './api';
import { Environment, Flag, FlagConfiguration } from './models';
import { FlagStore } from './flagStore';

const STRING_DELIMETERS = ['"', "'", '`'];
const FLAG_KEY_REGEX = /[A-Za-z0-9][\.A-Za-z_\-0-9]*/;
const LD_MODE: DocumentFilter = {
	scheme: 'file',
};

export function register(ctx: ExtensionContext, config: Configuration, flagStore: FlagStore, api: LaunchDarklyAPI) {
	ctx.subscriptions.push(
		commands.registerCommand('extension.configureLaunchDarkly', async () => {
			try {
				const configurationMenu = new ConfigurationMenu(config, api);
				await configurationMenu.configure();
				if (configurationMenu.didChangeAccessToken()) {
					await flagStore.reload();
				}
				window.showInformationMessage('LaunchDarkly configured successfully');
			} catch (err) {
				console.error(err);
				window.showErrorMessage('An unexpected error occured, please try again later.');
			}
		}),
	);

	ctx.subscriptions.push(
		languages.registerCompletionItemProvider(
			LD_MODE,
			new LaunchDarklyCompletionItemProvider(config, flagStore),
			"'",
			'"',
		),
	);

	ctx.subscriptions.push(languages.registerHoverProvider(LD_MODE, new LaunchDarklyHoverProvider(config, flagStore)));

	ctx.subscriptions.push(
		commands.registerTextEditorCommand('extension.openInLaunchDarkly', async editor => {
			let flagKey = editor.document.getText(
				editor.document.getWordRangeAtPosition(editor.selection.anchor, FLAG_KEY_REGEX),
			);
			if (!flagKey) {
				window.showErrorMessage(
					'[LaunchDarkly] Error retrieving flag (current cursor position is not a feature flag).',
				);
				return;
			}

			if (!config.accessToken) {
				window.showErrorMessage('[LaunchDarkly] accessToken is not set.');
				return;
			}

			if (!config.project) {
				window.showErrorMessage('[LaunchDarkly] project is not set.');
				return;
			}

			try {
				await openFlagInBrowser(config, flagKey, flagStore);
			} catch (err) {
				let errMsg = `Encountered an unexpected error retrieving the flag ${flagKey}`;
				if (err.statusCode == 404) {
					// Try resolving the flag key to kebab case
					try {
						await openFlagInBrowser(config, kebabCase(flagKey), flagStore);
						return;
					} catch (err) {
						if (err.statusCode == 404) {
							errMsg = `Could not find the flag ${flagKey}`;
						}
					}
				}
				console.error(err);
				window.showErrorMessage(`[LaunchDarkly] ${errMsg}`);
			}
		}),
	);
}

class LaunchDarklyHoverProvider implements HoverProvider {
	private readonly flagStore: FlagStore;
	private readonly config: Configuration;

	constructor(config: Configuration, flagStore: FlagStore) {
		this.config = config;
		this.flagStore = flagStore;
	}

	public provideHover(document: TextDocument, position: Position): Thenable<Hover> {
		return new Promise(async (resolve, reject) => {
			if (this.config.enableHover) {
				let candidate = document.getText(document.getWordRangeAtPosition(position, FLAG_KEY_REGEX));
				try {
					let data =
						(await this.flagStore.getFeatureFlag(candidate)) ||
						(await this.flagStore.getFeatureFlag(kebabCase(candidate)));
					if (data) {
						let hover = generateHoverString(data.flag, data.config);
						resolve(new Hover(hover));
						return;
					}
				} catch (e) {
					reject(e);
				}
			}
			reject();
		});
	}
}

class LaunchDarklyCompletionItemProvider implements CompletionItemProvider {
	private readonly flagStore: FlagStore;
	private readonly config: Configuration;

	constructor(config: Configuration, flagStore: FlagStore) {
		this.config = config;
		this.flagStore = flagStore;
	}

	public provideCompletionItems(document: TextDocument, position: Position): Thenable<CompletionItem[]> {
		if (isPrecedingCharStringDelimeter(document, position)) {
			return new Promise(async resolve => {
				if (this.config.enableAutocomplete) {
					const flags = await this.flagStore.allFlags();
					resolve(
						Object.keys(flags).map(flag => {
							return new CompletionItem(flag, CompletionItemKind.Field);
						}),
					);
					return;
				}
				resolve();
			});
		}
	}
}

const openFlagInBrowser = async (config: Configuration, flagKey: string, flagStore: FlagStore) => {
	const { flag } = await flagStore.getFeatureFlag(flagKey);

	// Default to first environment
	let env: Environment = Object.values(flag.environments)[0];
	let sitePath = env._site.href;

	if (!config.env) {
		window.showWarningMessage('[LaunchDarkly] env is not set. Falling back to first environment.');
	} else if (!flag.environments[config.env]) {
		window.showWarningMessage(
			`[LaunchDarkly] Configured environment '${config.env}' has been deleted. Falling back to first environment.`,
		);
	} else {
		env = flag.environments[config.env];
		sitePath = env._site.href;
	}
	opn(url.resolve(config.baseUri, sitePath));
};

export function generateHoverString(flag: Flag, c: FlagConfiguration) {
	return `**LaunchDarkly feature flag**\n
	Name: ${flag.name}
	Key: ${c.key}
	Enabled: ${c.on}
	Default variation: ${JSON.stringify(c.variations[c.fallthrough.variation])}
	Off variation: ${JSON.stringify(c.variations[c.offVariation])}
	${plural(c.prerequisites.length, 'prerequisite', 'prerequisites')}
	${plural(
		c.targets.reduce((acc, curr) => acc + curr.values.length, 0),
		'user target',
		'user targets',
	)}
	${plural(c.rules.length, 'rule', 'rules')}`;
}

function plural(count: number, singular: string, plural: string) {
	return count === 1 ? `1 ${singular}` : `${count} ${plural}`;
}

export function isPrecedingCharStringDelimeter(document: TextDocument, position: Position) {
	const range = document.getWordRangeAtPosition(position, FLAG_KEY_REGEX);
	if (!range || !range.start || range.start.character === 0) {
		return false;
	}
	const c = new Range(
		range.start.line,
		candidateTextStartLocation(range.start.character),
		range.start.line,
		range.start.character,
	);
	const candidate = document.getText(c).trim();
	return STRING_DELIMETERS.indexOf(candidate) !== -1;
}

const candidateTextStartLocation = (char: number) => (char === 1 ? 0 : char - 2);
