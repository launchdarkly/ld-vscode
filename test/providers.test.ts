import * as assert from 'assert';
import * as path from 'path';
import * as vscode from 'vscode';

import * as providers from '../src/providers';
import { Flag, FlagConfiguration } from '../src/models';

const flag = new Flag ({
	name: "Test",
	key: "test",
	tags: [],
	environments: null,
	_version: 1,
});

const flagConfig: FlagConfiguration = {
	key: 'test',
	on: true,
	variations: ['SomeVariation', { thisIsJson: 'AnotherVariation' }],
	fallthrough: {
		variation: 0,
	},
	offVariation: 1,
	prerequisites: ['something'],
	targets: [{ values: ['user', 'anotheruser'] }, { values: ['someotheruser'] }],
	rules: [],
	version: 1,
};

let testPath = path.join(__dirname, '..', '..', 'test');

suite('provider utils tests', () => {
	test('generateHoverString', () => {
		assert.equal(
			"**LaunchDarkly feature flag**\n\nName: \n```\nTest\n```\n\n\nKey: \n```\ntest\n```\n\n\nEnabled: \n```\ntrue\n```\n\n\nDefault variation: \n```\n\"SomeVariation\"\n```\n\n\nOff variation: \n```\n{\n  \"thisIsJson\": \"AnotherVariation\"\n}\n```\n\n\n1 prerequisite\n\n3 user targets\n\n0 rules",
			providers.generateHoverString(flag, flagConfig).value,
		);
	});

	test('isPrecedingCharStringDelimeter', async () => {
		// TODO: generate the test data in this file
		const uri = vscode.Uri.file(path.join(testPath, 'test.txt'));
		const tests = [
			{
				name: "single-quote delim",
				expected: true,
				line: 0,
				char: 1,
			},
			{
				name: "double-quote delim",
				expected: true,
				line: 1,
				char: 1,
			},
			{
				name: "backtick delim",
				expected: true,
				line: 2,
				char: 1,
			},
			{
				name: "not start of line",
				expected: true,
				line: 3,
				char: 2,
			},
			{
				name: "delim preceded by another char",
				expected: false,
				line: 4,
				char: 2,
			},
			{
				name: "doesn't match flag key regex",
				expected: false,
				line: 6,
				char: 2,
			}
		];

		const document = await vscode.workspace.openTextDocument(uri);
		tests.forEach(t => {
			const pos = new vscode.Position(t.line, t.char);
			assert.equal(providers.isPrecedingCharStringDelimeter(document, pos), t.expected, t.name);
		});
	});
});
