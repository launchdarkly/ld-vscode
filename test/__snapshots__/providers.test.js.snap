// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"![Flag status](data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTMiIGhlaWdodD0iMTMiIHZpZXdCb3g9IjAgMCAxMyAxMyIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZmlsbC1ydWxlPSJldmVub2RkIiBjbGlwLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0zLjI1IDRDMS40NTUwNyA0IDAgNS40NTUwNyAwIDcuMjVDMCA5LjA0NDkyIDEuNDU1MDcgMTAuNSAzLjI1IDEwLjVIOS43NUMxMS41NDQ5IDEwLjUgMTMgOS4wNDQ5MiAxMyA3LjI1QzEzIDUuNDU1MDcgMTEuNTQ0OSA0IDkuNzUgNEgzLjI1Wk05LjQ3OTE3IDkuNjg3NUMxMC44MjU0IDkuNjg3NSAxMS45MTY3IDguNTk2MTkgMTEuOTE2NyA3LjI1QzExLjkxNjcgNS45MDM4MSAxMC44MjU0IDQuODEyNSA5LjQ3OTE3IDQuODEyNUM4LjEzMjk3IDQuODEyNSA3LjA0MTY3IDUuOTAzODEgNy4wNDE2NyA3LjI1QzcuMDQxNjcgOC41OTYxOSA4LjEzMjk3IDkuNjg3NSA5LjQ3OTE3IDkuNjg3NVoiIGZpbGw9IiMyN0FFNjAiLz4KPC9zdmc+Cg==) abc / test / **[test](https://example.com/ \\"Open in LaunchDarkly\\")** 



First flag

* Prerequisites: \`something\`
* Targets configured


**$(symbol-boolean) Variations**

* \`1\` **one**: first flag \`$(arrow-small-right)fallthrough\`"
`;

exports[`provider utils tests generateHoverString 1`] = `
"![Flag status](data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTMiIGhlaWdodD0iMTMiIHZpZXdCb3g9IjAgMCAxMyAxMyIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHBhdGggZmlsbC1ydWxlPSJldmVub2RkIiBjbGlwLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0zLjI1IDRDMS40NTUwNyA0IDAgNS40NTUwNyAwIDcuMjVDMCA5LjA0NDkyIDEuNDU1MDcgMTAuNSAzLjI1IDEwLjVIOS43NUMxMS41NDQ5IDEwLjUgMTMgOS4wNDQ5MiAxMyA3LjI1QzEzIDUuNDU1MDcgMTEuNTQ0OSA0IDkuNzUgNEgzLjI1Wk05LjQ3OTE3IDkuNjg3NUMxMC44MjU0IDkuNjg3NSAxMS45MTY3IDguNTk2MTkgMTEuOTE2NyA3LjI1QzExLjkxNjcgNS45MDM4MSAxMC44MjU0IDQuODEyNSA5LjQ3OTE3IDQuODEyNUM4LjEzMjk3IDQuODEyNSA3LjA0MTY3IDUuOTAzODEgNy4wNDE2NyA3LjI1QzcuMDQxNjcgOC41OTYxOSA4LjEzMjk3IDkuNjg3NSA5LjQ3OTE3IDkuNjg3NVoiIGZpbGw9IiMyN0FFNjAiLz4KPC9zdmc+Cg==) abc / test / **[test](https://example.com/ \\"Open in LaunchDarkly\\")** 



First flag

* Prerequisites: \`something\`
* Targets configured


**$(symbol-boolean) Variations**

* \`1\` **one**: first flag \`$(arrow-small-right)fallthrough\`"
`;
